Redirect to DSC if enterprise offer applied with 100% discount

muhammad-ammar opened the pull request:
**Description:** During enrollment, if enterprise customer has enabled data sharing consent, the enterprise learner is not prompted for DSC if an enterprise offer with 100% discount applied to basket. This PR will redirect the learner to DSC page if enterprise customer required the consent.

**JIRA:** [ENT-2250](https://openedx.atlassian.net/browse/ENT-2250)

**Sandbox:** https://mammar.sandbox.edx.org/

**Testing:** This PR is deployed on sandbox.

**Test Data:** Below data exists on sandbox for testing
**verified course:** https://mammar.sandbox.edx.org/courses/course-v1:edX+JS101+2019_T1/about
**enterprise customer uuid:** cc8a0260-36d0-4520-8e1a-154d0431aa9c
**linked enterprise learners:** enterprise_learner_5XX@example.com to enterprise_learner_550@example.com

**NOTE:** **It would be great if reviewers/testers can pick different learner emails to avoid confusion**

**Testing Instructions:** One way to test the scenario could be like below
1. Login to LMS with any of the linked learners
2. Explorer the courses
3. Enroll into verified course
4. It takes the learner to payment MFE
5. Learner will be redirected to DSC page
6. If learner provides the consent then learner will be able to see the course content else it will be redirect to LMS dashboard and a consent declined notification will shown in the dashboard page.
7. If learner decline the consent and then go to course page, learner will again prompted for DSC.

georgebabey commented on ecommerce/enterprise/views.py:
```diff
@@ -89,3 +89,24 @@ def get_context_data(self, **kwargs):
+
+class EnterpriseConsentErrorView(TemplateView):
+    """
+    Display error message if enterprise learner choosed not to give data sharing consent for enterprise related offer.
```
nit: chose

irfanuddinahmad replied:
choosed -> chose

irfanuddinahmad commented on ecommerce/extensions/basket/views.py:
```diff
@@ -29,7 +30,14 @@
+    construct_enterprise_course_consent_url,
+    enterprise_customer_user_needs_consent,
+    get_enterprise_customer_from_voucher,
+    get_enterpruse_customer_from_enterprise_offer,
```
enterpruse -> enterprise

irfanuddinahmad commented:
Tested with enterprise_learner_502@example.com on sandbox ... works!
Great test setup ... good job!

muhammad-ammar commented:
@robrap Can you please take a look at this?

muhammad-ammar commented on ecommerce/enterprise/utils.py:
```diff
@@ -581,3 +585,53 @@ def get_enterprise_id_for_user(site, user):
+    )
+    success_url = '{base}{resource}'.format(
+        base=base_url,
+        resource=reverse('checkout:free-checkout'),
```
I will see if we can use `absolute_url` and above code suggested from https://github.com/edx/ecommerce/blob/7ec30a5d2a883cfee5b47bbd1ff66f908a19d0d3/ecommerce/extensions/basket/views.py#L821-L841

muhammad-ammar commented:
@robrap Thanks a lot for you thoughtful review. I agree that this whole flow is complicated but I was not able to find a better approach to fix the bug.

Please see my replies below.

We need `enterprise:consent_error` page to show the error message if learner declines to provide consent. I don't like adding this page but I am not sure if there is an other way to show the error message. I will check further if we can void adding this new template view.

---------

> @muhammad-ammar: I can't guarantee that this change I made didn't cause an issue, so not sure whether we should consider removing it:
> 
> https://github.com/edx/ecommerce/blob/dbdb4da3e64992391cd37d4baad3803ee544daf4/ecommerce/coupons/views.py#L147
> 
> To test, you need to ensure that you do **not** have an ecommerce session, which means not having the ecommerce-session cookie and being forced through the SSO flow if you go to a different ecommerce url without this permission class. In any case, if you are trying to remove variables, you could consider this, but it will slow down the enterprise flow because the user won't get to skip SSO.

Unfortunately I am not able to understand this, can you please explain? I am not sure how `CouponRedeemView` is related to changes made in this PR?

muhammad-ammar commented:
@robrap I have removed the `enterprise:consent_error` page and now in case consent decline, learner will be redirected to LMS dashboard and a consent declined notification will be shown to learner. This LMS notification logic is already present, we just need to pass the correct parameters in the redirect url. Other than this change everything else works as mentioned before. I think this approach is much better and cleaner than the previous. I have also addressed your `absolute_url` feedback.

Sandbox is updated with the latest changes.

Here is the LMS dashboard screenshot with consent decline notification
<img width="1439" alt="Screenshot 2019-10-08 at 5 11 18 PM" src="https://user-images.githubusercontent.com/6767924/66394492-b20a0480-e9ee-11e9-8ee5-1591a3c7c890.png">

muhammad-ammar commented:
@robrap I am happy too that we found a clean solution. Thanks for all the help.
Regarding from code reuse from below, I have checked and this part of the code is specifically written for voucher redemption.  
https://github.com/edx/ecommerce/blob/7ec30a5d2a883cfee5b47bbd1ff66f908a19d0d3/ecommerce/extensions/basket/views.py#L821-L841

The new implemented approach is not tested by others in the team, so I think we should wait for tomorrow or @georgebabey if you have some time then can you please test the flow on sandbox?

muhammad-ammar commented:
@robrap I have tested the changes myself and I don't think more testing is needed. Can you please merge the PR.
